File: hooks/go/hooks/custom_hook.go
package hooks

import (
	"fmt"
)

type CustomHook struct{}

func NewCustomHook() Hook {
	return &CustomHook{}
}

func (h *CustomHook) BeforeRequest(req Request) Request {
	fmt.Printf("Before request on URL %#v with method %#v\n", req.GetBaseUrl(), req.GetMethod())
	return req
}

func (h *CustomHook) AfterResponse(req Request, resp Response) Response {
	fmt.Printf("After response on URL %#v with method %#v, returning status %d\n", req.GetBaseUrl(), req.GetMethod(), resp.GetStatusCode())
	return resp
}

func (h *CustomHook) OnError(req Request, resp ErrorResponse) ErrorResponse {
	fmt.Printf("On Error: %#v\n", resp.Error())
	return resp
}

File: hooks/csharp/CustomHook.cs
#pragma warning disable 1998 // This async method lacks 'await' operators
#pragma warning disable 8603 // Possible null reference return.

public class CustomHook : IHook
{
  public async Task<HttpRequestMessage> BeforeRequestAsync(HttpRequestMessage request)
  {
    Console.WriteLine($"Before request on URL {request?.RequestUri?.AbsoluteUri} with method {request?.Method.ToString().ToUpper()}");
    return request;
  }

  public async Task<HttpResponseMessage> AfterResponseAsync(HttpResponseMessage response)
  {
    Console.WriteLine($"After response on URL {response?.RequestMessage?.RequestUri?.AbsoluteUri} with method {response?.RequestMessage?.Method.ToString().ToUpper()}, returning status {response?.StatusCode}");
    return response;
  }

  public async Task OnErrorAsync(HttpResponseMessage response)
  {
    Console.WriteLine($"On error - {response?.StatusCode} - {response?.ReasonPhrase}");
  }
}
